lib/bap_llvm/llvm_macho_loader.hpp

```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+        if (rel && obj.getPlainRelocationExternal(*rel))
+            symbol_reference(obj, obj.getPlainRelocationSymbolNum(*rel), rel->r_word0, s);
     }
 }
 
-void indirect_symbols(const macho &obj, const MachO::dysymtab_command &dlc, ogre_doc &s) {
+symbol_iterator get_indirect_symbol(const macho &obj, const MachO::dysymtab_command &dlc, uint32_t index) {
+    uint64_t obj_size = obj.getData().size();
 
-    for (auto sec : prim::sections(obj)) { // sections
-        auto typ = section_type(obj, sec);
-        if (typ == MachO::S_NON_LAZY_SYMBOL_POINTERS ||
-            typ == MachO::S_LAZY_SYMBOL_POINTERS ||
-            typ == MachO::S_LAZY_DYLIB_SYMBOL_POINTERS ||
-            typ == MachO::S_THREAD_LOCAL_VARIABLE_POINTERS ||
-            typ == MachO::S_SYMBOL_STUBS) { // section test
-
-            uint32_t stride;
-            if (typ == MachO::S_SYMBOL_STUBS)
-                stride = section_reserved2(obj, sec);
-            else
-                stride = obj.is64Bit() ? 8 : 4;
-            if (stride == 0) continue;
-
-            uint64_t count = section_size(obj, sec) / stride;
-            uint32_t n = section_reserved1(obj, sec);
-            uint64_t addr = section_addr(obj, sec);
-            uint32_t offs = section_offset(obj, sec);
-
-            for (uint32_t j = 0; j < count && n + j < dlc.nindirectsyms; j++) {
-                auto i_addr = addr + j * stride;
-                auto i_offs = offs + j * stride;
-                uint32_t indirect_symbol = obj.getIndirectSymbolTableEntry(dlc, n + j);
-                MachO::symtab_command Symtab = obj.getSymtabLoadCommand();
-                if (indirect_symbol < Symtab.nsyms) {
-                    symbol_iterator Sym = get_symbol(obj, indirect_symbol);
-
-                    if (auto name = prim::symbol_name(*Sym)) {
-                        s.entry("symbol-entry") << *name << i_addr << stride << i_offs ;
-                        s.entry("code-entry") << *name << i_offs << stride;
+    // this is actually a check that we will not get out of object
+    // bounds while calling getIndirectSymbolTableEntry.
+    // llvm does approximately the same to raise a nasty
+    // runtime LLVM_ERROR
+    uint64_t offset = dlc.indirectsymoff + index * sizeof(uint32_t);
+    symbol_iterator s = prim::end_symbols(obj);
+    if (offset < obj_size)
+        s = get_symbol(obj, obj.getIndirectSymbolTableEntry(dlc, index));
+    return s;
+}
+
+// returns stride for indirect symbols entries
+uint32_t indirect_symbols_stride(const macho &obj, const SectionRef &sec) {
+    auto typ = section_type(obj, sec);
+    uint32_t stride = 0;
+    if (typ == MachO::S_SYMBOL_STUBS)
+        stride = section_reserved2(obj, sec);
+    else
+        stride = obj.is64Bit() ? 8 : 4;
+    return stride;
+}
+
+// true for symbol stubs or symbol pointers sections
+bool contains_sym_stubs(const macho &obj, const SectionRef &sec) {
+    auto typ = section_type(obj, sec);
+    return
```

Comment:
you do not need to put parentheses around the return expression. You can also use switch/case

```c++
switch (section_type(obj, sec)) {
  case X:
  case Y:  
  case Z:   return true;
  default:  return false;
}
```

That's is very close to OCaml's match.